lib/sequelize.js

```diff
@@ -974,41 +975,61 @@ class Sequelize {
     // testhint argsConform.end
 
     const transaction = new Transaction(this, options);
-    const ns = Sequelize.cls;
-
-    if (autoCallback) {
-      let transactionResolver = (resolve, reject) => {
-        transaction.prepareEnvironment().then(() => {
-          if (ns) {
-            autoCallback = ns.bind(autoCallback);
-          }
-
-          const returnValue = autoCallback(transaction);
-          if (!returnValue || !returnValue.then) throw new Error('You need to return a promise chain/thenable to the sequelize.transaction() callback');
-
-          return returnValue.then(result => transaction.commit()).then(() => {
-            resolve(returnValue);
-          });
-        }).catch(err => {
-          // If the transaction has already finished (commit, rollback, etc), reject with the original error
-          if (transaction.finished) {
-            reject(err);
-          } else {
-            return transaction.rollback().finally(() => {
-              reject(err);
-            });
-          }
-        });
-      };
-
-      if (ns) {
-        transactionResolver = ns.bind(transactionResolver, ns.createContext());
-      }
+    const ns = Sequelize._cls;
+
+    if (!autoCallback) return transaction.prepareEnvironment().return(transaction);
+
+    // autoCallback provided
+    const wrapper = ns ? function(fn) {
+      var promise;
+      ns.run(() => promise = fn());
+      return promise;
+    } : function(fn) { return fn(); };
+
+    return wrapper(() =>
+      transaction.prepareEnvironment().then(() =>
+        autoCallback(transaction)
+      ).tap(() =>
+        transaction.commit()
+      ).catch(err =>
+        // Rollback transaction if not already finished (commit, rollback, etc) and reject with original error
+        Promise.try(() => {
+          // Rollback (ignore any error in rollback)
+          if (!transaction.finished) return transaction.rollback().catch(function() {});
+        }).throw(err)
+      )
+    );
+  }
 
-      return new Promise(transactionResolver);
-    } else {
-      return transaction.prepareEnvironment().return(transaction);
-    }
+  /**
+   * Use CLS with Sequelize.
+   * CLS namespace provided is stored as `Sequelize._cls`
+   * and bluebird Promise is patched to use the namespace, using `cls-bluebird` module.
+   *
+   * @param {Object}   ns   CLS namespace
+   * @returns {Object}      Sequelize constructor
+   */
+  static useCls(ns) {
+    // check `ns` is valid CLS namespace
+    if (!ns || typeof ns !== 'object' || typeof ns.bind !== 'function' || typeof ns.run !== 'function') throw new Error('Must provide CLS namespace');
+
+    // save namespace as `Sequelize._cls`
+    this._cls = ns;
+
+    // patch bluebird to bind all promise callbacks to CLS namespace
+    clsBluebird(ns, Promise);
+
+    // return Sequelize for chaining
+    return this;
+  }
+
+  static get cls() {
+    return this._cls;
+  }
```

Comment:
Idea is that `Sequelize.cls` should not be used. But I put the getter/setter in there so as not to break compatibility. If you do `Sequelize.cls = ns` rather than `Sequelize.useCls(ns)` it works, but gives a deprecation warning.

I thought it was nice to maintain the old API for this version (but deprecated) rather than just strip it out entirely.

I could take it out, or also add a deprecation warning to the getter.

@mickhansen `Sequelize._cls` is used everywhere internally in Sequelize, so the extra overhead of the getter should never be incurred.